source/common/ssl/context_impl.cc

```diff
@@ -648,54 +654,58 @@ ServerContextImpl::ServerContextImpl(Stats::Scope& scope, const ServerContextCon
           });
     }
 
-    setSessionIdContext(ctx, server_names);
+    int rc = SSL_CTX_set_session_id_context(ctx.ssl_ctx_.get(), session_context_buf,
+                                            session_context_len);
+    RELEASE_ASSERT(rc == 1, "");
   }
 }
 
-void ServerContextImpl::setSessionIdContext(TlsContext& ctx,
-                                            const std::vector<std::string>& server_names) {
-  X509* cert = SSL_CTX_get0_certificate(ctx.ssl_ctx_.get());
-  uint8_t session_context_buf[EVP_MAX_MD_SIZE] = {};
-  unsigned session_context_len = 0;
+void ServerContextImpl::hashSessionContextId(const std::vector<std::string>& server_names,
+                                             uint8_t* session_context_buf,
+                                             unsigned& session_context_len) {
   EVP_MD_CTX md;
   int rc = EVP_DigestInit(&md, EVP_sha256());
   RELEASE_ASSERT(rc == 1, "");
 
-  // Hash the CommonName/SANs of the server certificate. This makes sure that sessions can only be
-  // resumed to a certificate for the same name, but allows resuming to unique certs in the case
-  // that different Envoy instances each have their own certs.
-  X509_NAME* cert_subject = X509_get_subject_name(cert);
-  RELEASE_ASSERT(cert_subject != nullptr, "");
-  int cn_index = X509_NAME_get_index_by_NID(cert_subject, NID_commonName, -1);
-  // It's possible that the certificate doesn't have CommonName, but has SANs.
-  if (cn_index >= 0) {
-    X509_NAME_ENTRY* cn_entry = X509_NAME_get_entry(cert_subject, cn_index);
-    RELEASE_ASSERT(cn_entry != nullptr, "");
-    ASN1_STRING* cn_asn1 = X509_NAME_ENTRY_get_data(cn_entry);
-    RELEASE_ASSERT(ASN1_STRING_length(cn_asn1) > 0, "");
-    rc = EVP_DigestUpdate(&md, ASN1_STRING_data(cn_asn1), ASN1_STRING_length(cn_asn1));
-    RELEASE_ASSERT(rc == 1, "");
-  }
+  for (const auto& ctx : tls_contexts_) {
+    X509* cert = SSL_CTX_get0_certificate(ctx.ssl_ctx_.get());
+    // Hash the CommonName/SANs of the server certificate. This makes sure that sessions can only be
```

Comment:
Nit: `the server certificate` -> `all server certificates` (and reword the whole comment, perhaps even move it outside the loop).